File: leetcode-python-go/977. 有序数组的平方/main2.go
package main

import (
	"fmt"
)

func sortedSquares(nums []int) []int {
	n := len(nums)
	ans := make([]int, n)
	i, j := 0, n-1
	for pos := n - 1; pos >= 0; pos-- {
		if v, w := nums[i]*nums[i], nums[j]*nums[j]; v > w {
			ans[pos] = v
			i++
		} else {
			ans[pos] = w
			j--
		}
	}
	return ans
}

func main() {
	nums := []int{-4, -1, 0, 3, 10}
	results := sortedSquares(nums)
	fmt.Println(results)
}

File: leetcode-python-go/977. 有序数组的平方/main2.py
from typing import List

nums = [-4,-1,0,3,10]


def sortedSquares(nums: List[int]) -> List[int]:
    n = len(nums)
    ans = [0] * n
    
    i, j, pos = 0, n - 1, n - 1
    while i <= j:
        if nums[i] * nums[i] > nums[j] * nums[j]:
            ans[pos] = nums[i] * nums[i]
            i += 1
        else:
            ans[pos] = nums[j] * nums[j]
            j -= 1
        pos -= 1
    
    return ans

print(sortedSquares(nums))